Davem/xspod

iabyn opened the pull request:
Rewrite perlxs.pod

This branch completely rewrites and modernises the XS reference manual,
perlxs.pod.

The new file is about twice the size of the old one.

This branch:

- deletes some obsolete sections;

- reorders the existing sections into a more logical order;

- adds a large new introductory/overview part, which explains
  all the background needed to understand what XSUBs do, including
  SVs, the stack, reference counts, magic etc.

- includes a BNF syntax section

- modernises: e.g. it uses "ANSI" parameter syntax throughout

- has a fully-worked example using T_PTROBJ

Note that although each commit in this branch may have a complex-looking
diff for the updating of a particular section,  in reality most sections
haver been rewritten from scratch, and the diff output is showing
paragraph breaks as fixed unchanging points, so that it appears as lots of
individual paragraph changes rather than "delete all this text, add new
text". If reviewing, it may be easier to just read the final perlxs.pod
file instead of looking at the diffs.


* This set of changes requires a perldelta entry, and I will write one later

iabyn committed:
perlxs.pod: remove section on %v

The XS parser supported an extremely obscure bit of functionality which
made use of the %v package variable to maintain state between different
bits of typemap processing. This was accidentally broken in 5.10.0:
refactoring removed the 'use vars "%v"' line, and no one seemed to
notice or care.

Also, the sole example of its use in the docs seemed to be obscure,
confusing and probably wrong.

There was a consensus in the discussion at

    http://nntp.perl.org/group/perl.perl5.porters/267667

that we should stop documenting this feature rather than trying to fix
it.

iabyn committed:
perlxs.pod: reindent and reformat code examples

The various XS code examples had odd and inconsistent indentation (often
with 5 leading spaces) and inconsistent formatting, e.g. foo(a,b) vs
foo( a, b ) vs foo(a, b). Fix that, and also remove any tab chars.

Whitespace-only change.

iabyn committed:
perlxs.pod: delete most non-ref sections

This commit is a simple cut which deletes several '=head2' sections from
perlxs.pod. The next commit will tidy up and fix any broken links etc.

These sections are more tutorial-like, and aren't in line with the goal
of this branch that perlxs.pod becomes purely a reference manual for XS.
Any relevant information from these sections may be incorporated later
into new sections in perlxs.pod and/or be included in a future rewrite
of perlxstut.pod.

The sections deleted are:

    =head2 Introduction
    =head2 On The Road
    =head2 The Anatomy of an XSUB
    =head2 The Argument Stack
    =head2 The RETVAL Variable
    =head2 Returning SVs, AVs and HVs through RETVAL
    =head2 Returning Undef And Empty Lists
    =head2 Interface Strategy
    =head2 Perl Objects And C Structures

iabyn committed:
perlxs.pod: fix up links after deleting sections

The previous commit deleted several sections from perlxs.pod. This
commit fixes things up; done as a separate commit so that the changes
aren't drowned out in the diff listing.

iabyn committed:
perlxs.pod: reorder sections

This big commit does a series of plain cut+pastes to reorder all the
=head2 sections within the file.

This changes the order from semi-random into roughly the order the
various XS keywords would appear within an XS file, and then within an
XSUB declaration/definition.

No changes have been made to the text: simply that all lines from a
particular '^=head2' up until the next head2 have been cut+paste as
a single unit.

No attempt has been made yet to make the text consistent with the new
ordering; that will be done by the subsequent commits of this branch.

The previous ordering in this file was:

    =head1 NAME
    =head1 DESCRIPTION
    =head2 The MODULE Keyword
    =head2 The PACKAGE Keyword
    =head2 The PREFIX Keyword
    =head2 The OUTPUT: Keyword
    =head2 The NO_OUTPUT Keyword
    =head2 The CODE: Keyword
    =head2 The INIT: Keyword
    =head2 The NO_INIT Keyword
    =head2 The TYPEMAP: Keyword
    =head2 Initializing Function Parameters
    =head2 Default Parameter Values
    =head2 The PREINIT: Keyword
    =head2 The SCOPE: Keyword
    =head2 The INPUT: Keyword
    =head2 The IN/OUTLIST/IN_OUTLIST/OUT/IN_OUT Keywords
    =head2 The C<length(NAME)> Keyword
    =head2 Variable-length Parameter Lists
    =head2 The C_ARGS: Keyword
    =head2 The PPCODE: Keyword
    =head2 The REQUIRE: Keyword
    =head2 The CLEANUP: Keyword
    =head2 The POSTCALL: Keyword
    =head2 The BOOT: Keyword
    =head2 The VERSIONCHECK: Keyword
    =head2 The PROTOTYPES: Keyword
    =head2 The PROTOTYPE: Keyword
    =head2 The ALIAS: Keyword
    =head2 The OVERLOAD: Keyword
    =head2 The FALLBACK: Keyword
    =head2 The INTERFACE: Keyword
    =head2 The INTERFACE_MACRO: Keyword
    =head2 The INCLUDE: Keyword
    =head2 The INCLUDE_COMMAND: Keyword
    =head2 The CASE: Keyword
    =head2 The EXPORT_XSUB_SYMBOLS: Keyword
    =head2 The & Unary Operator
    =head2 Inserting POD, Comments and C Preprocessor Directives
    =head2 Using XS With C++
    =head2 Safely Storing Static Data in XS
    =head3 MY_CXT REFERENCE
    =head1 EXAMPLES
    =head1 CAVEATS
    =head2 Use of standard C library functions
    =head2 Event loops and control flow
    =head1 XS VERSION
    =head1 AUTHOR DIAGNOSTICS
    =head1 AUTHOR

and is now:

    =head1 NAME
    =head1 DESCRIPTION
    =head2 The MODULE Keyword
    =head2 The PACKAGE Keyword
    =head2 The PREFIX Keyword
    =head2 Inserting POD, Comments and C Preprocessor Directives
    =head2 The REQUIRE: Keyword
    =head2 The VERSIONCHECK: Keyword
    =head2 The PROTOTYPES: Keyword
    =head2 The EXPORT_XSUB_SYMBOLS: Keyword
    =head2 The INCLUDE: Keyword
    =head2 The INCLUDE_COMMAND: Keyword
    =head2 The TYPEMAP: Keyword
    =head2 The BOOT: Keyword
    =head2 The FALLBACK: Keyword
    =head2 The NO_OUTPUT Keyword
    =head2 The IN/OUTLIST/IN_OUTLIST/OUT/IN_OUT Keywords
    =head2 Default Parameter Values
    =head2 The C<length(NAME)> Keyword
    =head2 Variable-length Parameter Lists
    =head2 The PREINIT: Keyword
    =head2 The INPUT: Keyword
    =head2 The NO_INIT Keyword
    =head2 Initializing Function Parameters
    =head2 The & Unary Operator
    =head2 The SCOPE: Keyword
    =head2 The INIT: Keyword
    =head2 The C_ARGS: Keyword
    =head2 The CODE: Keyword
    =head2 The PPCODE: Keyword
    =head2 The POSTCALL: Keyword
    =head2 The OUTPUT: Keyword
    =head2 The CLEANUP: Keyword
    =head2 The PROTOTYPE: Keyword
    =head2 The OVERLOAD: Keyword
    =head2 The ALIAS: Keyword
    =head2 The INTERFACE: Keyword
    =head2 The INTERFACE_MACRO: Keyword
    =head2 The CASE: Keyword
    =head2 Using XS With C++
    =head2 Safely Storing Static Data in XS
    =head3 MY_CXT REFERENCE
    =head1 EXAMPLES
    =head1 CAVEATS
    =head2 Use of standard C library functions
    =head2 Event loops and control flow
    =head1 XS VERSION
    =head1 AUTHOR DIAGNOSTICS
    =head1 AUTHOR

iabyn committed:
perlxs.pod: add group headers

Following the previous commit's reordering of the all the =head2
sections, demote most of the =head2 headers to =head3, and add some new
=head2 headers which group together related headers.

Also add some =head3's for a few missing keywords.

Subsequent commits will flesh out the new sections.

iabyn committed:
perlxs.pod: add a new introductory part

Four commits ago, I removed most of the general text sections in
perlxs (i.e. the ones not specifically about a particular keyword).

Now this commit adds a completely new introductory part to perlxs, about
1200 lines long. It represents an attempt to write a background to what
XS and XSUBs, SVs, typemaps etc are, in a complete and modern way.
The existing reference section for each keyword follows it.

I tried to avoid getting too tutorial-like (that's what perlxstut is
for), but I may have crossed the line in various places. In particular
it has a new section which could have been titled "all the bits of
perlguts you need to know in order to write non-trivial XSUBs without
having to actually read perlguts".

iabyn committed:
perlxs.pod: add BNF definition section

Add a section which semi-formally tries to define the syntax and
structue of an XS file, using a BNF-like format.

See http://nntp.perl.org/group/perl.perl5.porters/268701 for the
discussion of this part.

iabyn committed:
perlxs.pod: update MODULE/PACKAGE/PREFIX

Rewrite the POD for these three keywords, and in particular, treat
them as one declaration, rather than three unrelated keywords.

iabyn committed:
perlxs.pod: update REQUIRE, VERSIONCHECK keywords

iabyn committed:
perlxs.pod: update PROTOTYPES: keyword

iabyn committed:
perlxs.pod: update EXPORT_XSUB_SYMBOLS, INCLUDE(_COMMAND)

iabyn committed:
perlxs.pod: update TYPEMAP: keyword

iabyn committed:
perlxs.pod: update BOOT: keyword

iabyn committed:
perlxs.pod: update FALLBACK: keyword

iabyn committed:
perlxs.pod: update XSUB Structure + Declaration

Populate the new

    =head2 The Structure of an XSUB
    =head2 An XSUB Declaration

sections

iabyn committed:
perlxs.pod: update file-scoped directive text

Populate the new

     =head2 File-scoped XS Keywords and Directives

section, partially by cannibalising (and then deleting) the old

    =head3 Inserting POD, Comments and C Preprocessor Directives

subsection. This commit only adds text about directives; subsequent
commits will update the various file-scoped keywords.

iabyn committed:
perlxs.pod: update section 'An XSUB Parameter'

Add some initial text for this new section, and also add a new
subsection "XSUB Parameter Placeholders".

iabyn committed:
perlxs.pod: update IN_OUT etc section

iabyn committed:
perlxs.pod: update default, length, ellipis params

Rewrite (and retitle) these three subsections:

    =head3 Default Parameter Values
    =head3 The C<length(NAME)> Keyword
    =head3 Variable-length Parameter Lists

iabyn committed:
perlxs.pod: update: Input part, PREINIT sections

Add text for the new '=head2 The XSUB Input Part' section, and rewrite
the existing entry for the PREINIT keyword.

iabyn committed:
perlxs.pod: update 'The INPUT: Keyword' section

This commit completely rewrites this section and subsections:

    =head3 The INPUT: Keyword
        =head4 The NO_INIT Keyword
        =head4 Initializing Function Parameters
        =head4 The & Unary Operator

It de-emphasises the INPUT keyword and suggests using ANSI XS signatures
etc instead.

iabyn committed:
perlxs.pod: update 'SCOPE: Keyword' section

iabyn committed:
perlxs.pod: update: init part, INIT sections

Add text for the new '=head2 The XSUB Init Part' section, and rewrite
the existing entry for the INIT keyword.

iabyn committed:
perlxs.pod: update: code part, autocall, C_ARGS

Add text to the new

    =head2 The XSUB Code Part
    =head3 Auto-calling a C function

sections, and rewrite the existing

    =head4 The C_ARGS: Keyword

section

iabyn committed:
perlxs.pod: update: CODE, PPCODE

Rewrite these sections:

    =head3 The CODE: Keyword
    =head3 The PPCODE: Keyword

iabyn committed:
perlxs.pod: update NOT_IMPLEMENTED_YET: keyword

This keyword formerly wasn't documented. The docs now say "this is what
it is, but don't use it".

iabyn committed:
perlxs.pod: update: output part

Add text to the new

    =head2 The XSUB Output Part

section, and rewrite the text in these existing sections:

    =head3 The POSTCALL: Keyword
    =head3 The OUTPUT: Keyword

iabyn committed:
perlxs.pod: update: cleanup part

Add text to the new

    =head2 The XSUB Cleanup Part

section, and rewrite the text in this existing section:

 =head3 The CLEANUP: Keyword

iabyn committed:
perlxs.pod: update generic intro, PROTOTYPE

Add text to the new

    =head2 XSUB Generic Keywords

section, and rewrite the text in this existing section:

    =head3 The PROTOTYPE: Keyword

iabyn committed:
perlxs.pod: update OVERLOAD, add T_PTROBJ

First, add a new subsection

    =head3 T_PTROBJ and opaque handles

to the TYPEMAPs section explaining how this typemap can be used to
map between Perl objects and C library handles. It provides a
fully-worked example of wrapping a simple arithmetic library.

Then completely rewrite the

    =head3 The OVERLOAD: Keyword

section. In particular, it now refers to the new T_PTROBJ example and
shows how it can be extended to use overloading.

iabyn committed:
perlxs.pod: mention package name types

Explain that a 'C' parameter type in an XSUB declaration can actually
be a Perl package name or similar, e.g.

    Foo::Bar
    f(Foo::Bar obj, char *s)

iabyn committed:
perlxs.pod: document ATTRS

This keyword was undocumented, even though it had been added 25 years
ago.

iabyn committed:
perlxs.pod: add "Sharing XSUB bodies" section

Populate the introduction to this new section.

iabyn committed:
perlxs.pod: update: ALIAS

Rewrite this section:

    =head3 The ALIAS: Keyword

iabyn committed:
perlxs.pod: update INTERFACE, INTERFACE_MACRO

Rewrite these sections:

 =head3 The INTERFACE: Keyword
 =head3 The INTERFACE_MACRO: Keyword

also demote the second to be a head4 child of the first. Then expand
the T_PTROBJ example to use INTERFACE as an alternative to ALIAS.

iabyn committed:
perlxs.pod: update: CASE

Rewrite this section:

    =head3 The CASE: Keyword

iabyn committed:
perlxs.pod: add "Using Typemaps" section

Populate this new section (except for the T_PTROBJ subsection, which had
already been added by an earlier commit within this branch).

Note that the "Common typemaps" subsection could probably benefit
from some further expansion by someone familiar with which built-in
T_FOO entries are useful.

iabyn committed:
perlxs.pod: update "Using XS With C++" section

Rewrite this section:

 =head2 Using XS With C++

Disclaimer: I've never written a proper C++ program. I had to
(literally) dust off my 34-year old copy of Stroustrup(*) and also do
some Googling. Hopefully what I've written is sane.

(*) This was bought back in the days when people used to to learn things
by buying books, and when I thought that I ought to know something about
this newfangled C++ thing. I never got round to reading all of it: I
discovered Perl around the same time, which looked to be a lot more fun.

iabyn committed:
perlxs.pod: update MY_CXT section

Revise the text in this section:

    =head2 Safely Storing Static Data in XS

iabyn committed:
perlxs.pod: update EXAMPLES section

Rewrite this section:

     =head1 EXAMPLES

Basically, delete the one big example in this section and instead
provide links to various other examples already present in this document
instead.

iabyn committed:
perlxs.pod: update CAVEATS, AUTHOR, A DIAGNOSTICS

Tweak the final few sections of perlxs.pod.

tonycoz commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+=item *
+
+L<perlcall>: this describes how to call Perl functions and do the
+equivalent of C<eval ""> from C.
+
+=item *
+
+L<perlembed>: this describes how to embed a complete Perl interpreter
+within another application.
```
perlcall barely mentions eval_pv, perlembed provides much better documentation for eval_pv/eval_sv.

tonycoz commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+There is a standard system typemap file which contains rules for common C
+and Perl types, but you can add your own typemap file in addition, and
+from perl 5.16.0 onwards you can also add typemap declarations inline
```
Shouldn't this refer to the version of EU::PXS that added embedded typemaps (3.01) instead of the perl version?

johannessen replied:
Practice throughout the document is inconsistent. There are numerous cases of “since 5.\*” or “Perl 5.\*”, but also some cases where both the Perl version and the ParseXS version are stated, e. g. in the sections [SCOPE](https://github.com/Perl/perl5/blob/9afbfdafc5e26a350cb4c430f5d881e35e598f43/dist/ExtUtils-ParseXS/lib/perlxs.pod#the-scope-keyword-and-typemap-entry) and [ALIAS](https://github.com/Perl/perl5/blob/9afbfdafc5e26a350cb4c430f5d881e35e598f43/dist/ExtUtils-ParseXS/lib/perlxs.pod#the-alias-keyword). And there’s at least one case where *only* the ParseXS version is stated (I forgot in which section).

I think trying to consistently state both versions might be useful: The ParseXS version because it’s technically the more relevant one here, and the Perl version because it’s more intuitive for many people.

Leont replied:
> I think trying to consistently state both versions might be useful: The ParseXS version because it’s technically the more relevant one here, and the Perl version because it’s more intuitive for many people.

No, I believe adding Perl versions will only add confusion here.

tonycoz commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+non-overloaded reference to an array may return a temporary string like
+C<"ARRAY(0x12345678)">.
+
+If you need to coerce an SV to a string (e.g. before modifying its string
+value) then use C<SvPV_force()> or one of its variants. For example if
```
Perhaps "before directly modifying its string buffer".

You don't need to force normal before calling sv_setpvn().

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+in a particular C library, the XSUB definitions in the XS file are often
+just a couple of lines, consisting of a declaration of the name,
+parameters and return type. The XS parser will do almost all the heavy
+lifting for you,
```
```perl
s/you,/you./
```

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+    package Foo::Bar;
+    our $VERSION = '1.01';
+    require XSLoader;
+    XSLoader::load(__PACKAGE__, $VERSION);
```
XSLoader 0.14 (core since Perl 5.15) offers a [simpler syntax](https://metacpan.org/release/SAPER/XSLoader-0.24/view/XSLoader.pm):
```perl
XSLoader::load();
```
Given that the goal of this rewrite is to modernise the documentation, should this newer syntax be used in the example?

Leont replied:
That automatically adds the module name but not the expected version. Personally I wouldn't use it..

mauke replied:
`perldoc XSLoader` says:

> A sanity check is done to ensure that the versions of the .pm and the (compiled) .xs parts are compatible. If `$VERSION` was specified, this is used for the check. If not specified, it defaults to `$XS_VERSION // $VERSION` (in the module's namespace)

So that sounds fine to me.

johannessen replied:
Right. The XSLoader documentation is correct. The default is applied by [S_xs_version_bootcheck()](https://github.com/Perl/perl5/blob/87a16f665d01ed72959b9a366e8ab74a97f9cb4b/util.c#L5598-L5608), which is called during the handshake.

I think the overall recommendation for users should be to use no arguments in `XSLoader::load()`, for simplicity. So if there’s no particular strategy behind showing these arguments in the example code, I think they should be removed here.

Leont replied:
You're right. It's just implemented completely differently from what I expected so I erroneously concluded that that didn't work.

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+Known C</^#/> C preprocessor directives, which are passed through as-is.
+Conditional ones, such as C<#if> and C<#else>, have some basic analysis
+performed on them which, in particular, allows two variants of the same
+XSUB to be declared without raising a "duplicate XSUB" warning. This
+warning suppression only works for the if/else/endif form. For example
+this works:
```
I was, at first, confused by this text: From reading it, and the examples below it, it seemed like what *doesn’t* work is the `#ifndef` directive, specifically. Furthermore, the examples use “ifdef”, whereas this text uses “if”, which made it look like there’s a typo here.

After re-reading the [old perlxs document](https://perldoc.perl.org/5.42.0/perlxs#Inserting-POD,-Comments-and-C-Preprocessor-Directives), I understand presence of `#else` in particular is the key here. I think the new text isn’t very clear if you don’t already know that. How about something like this:

```
XSUB to be declared without raising a "duplicate XSUB" warning. This
warning suppression only works when an else branch is present. For
example, this works:
```

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+The C<REQUIRE> keyword is used to indicate the minimum version of the
+C<ExtUtils::ParseXS> XS compiler (and its F<xsubpp> wrapper) needed to
+compile the XS module. It is expected to be a floating-point number of the
+form C<\d+\.\d+/>. It is analogous to the perl C<use v5.xx>.
```
```patch
-form C<\d+\.\d+/>. It is analogous to the perl C<use v5.xx>.
+form C<\d+\.\d+/>. It is analogous to the perl C<use ExtUtils::ParseXS x.xx>.
```

tonycoz replied:
Also, the version regex has a trailing / but not a leading /.

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+The name of the XSUB is usually put on the line following the type, in
+which case it must be on column one. It is permissible for both the return
+type and name to be on the same line.
```
Very minor nit: I think those sentences would read better if the phrase “return type” was used first.

```patch
-The name of the XSUB is usually put on the line following the type, in
-which case it must be on column one. It is permissible for both the return
+The name of the XSUB is usually put on the line following the return type,
+in which case it must be on column one. It is permissible for both the
 type and name to be on the same line.
```

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+    Usage: Foo::Bar::foo(a, SV*, c) at ...
+
+Placeholders can't be used with autocall unless you use C<C_ARGS> to
+override the missing argument. For example:
```
The `C_ARGS` keyword is introduced somewhat suddenly here. I would liked to have seen a cross-reference:

```
Placeholders can't be used with autocall unless you use L<C_ARGS|/The
C_ARGS: Keyword> to override the missing argument. For example:
```

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+    foo(int a, b, char *c)
+        C_ARGS: a, c
+
+=head3 Updating and returning parameter values: the IN_OUT etc keywords
+
+               int i
+    IN         int i
+    IN_OUT     int i
+    IN_OUTLIST int i
+    OUT        int i
+    OUTLIST    int i
```
The pod spec doesn’t seem to specify the semantics of verbatim paragraphs where the first line is indented farther than other lines. Some pod parsers (notably GitHub’s) don’t deal well with such cases, [messing up the indentation](https://github.com/Perl/perl5/blob/9afbfdafc5e26a350cb4c430f5d881e35e598f43/dist/ExtUtils-ParseXS/lib/perlxs.pod#updating-and-returning-parameter-values-the-in_out-etc-keywords).

Listing `IN` before the default case would avoid the issue:
```patch
+    IN         int i
                int i
-    IN         int i
     IN_OUT     int i
```

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+The text of the modifier has up to two parts, separated by an underscore.
+The input part comes first and can have the value C<''> or C<IN>, while
+the output part can be one of C<''>, C<OUT>, or C<OUTLIST>.
```
According to this, `IN_''` would be a valid modifier. Or, if you assume that `''` signifies the empty string, `IN_` would be a valid modifier.

Yes, if you already know how this works, it should be reasonably clear. But I, for one, found it confusing in context even after reading it multiple times. How about something along the lines of this:

```patch
@@ -2260,7 +2260,7 @@
 =item *
 
-The text of the modifier has up to two parts, separated by an underscore.
-The input part comes first and can have the value C<''> or C<IN>, while
-the output part can be one of C<''>, C<OUT>, or C<OUTLIST>.
+The text of the modifier has one or two parts, separated by an underscore.
+The input part comes first and, if present, has the value C<IN>; while
+the output part, if present, can be one of C<OUT> or C<OUTLIST>.
 
 =item *
@@ -2275,5 +2275,5 @@
 =item *
 
-An output part of C<''> (the default) means nothing is done to return the
+An absent output part (the default) means nothing is done to return the
 value of the auto variable.
 
```

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+=back
+
+The approximate perl equivalents for these modifiers are given in the
+examples below, where the perl code C<real_foo(\$i)> stands in for the C
+autocall C<foo(&i)>.
+
+
+       int i             sub foo {
+    IN int i                  my $i = $_[N];
+                              real_foo($i);
+                         }
```
The pod spec doesn’t seem to specify the semantics of verbatim paragraphs where the first line is indented farther than other lines. Some pod parsers (notably GitHub’s) don’t deal well with such cases, [messing up the indentation](https://github.com/Perl/perl5/blob/9afbfdafc5e26a350cb4c430f5d881e35e598f43/dist/ExtUtils-ParseXS/lib/perlxs.pod#updating-and-returning-parameter-values-the-in_out-etc-keywords).

Listing `IN` before the default case would avoid the issue:
```patch
-       int i             sub foo {
-    IN int i                  my $i = $_[N];
+    IN int i             sub foo {
+       int i                  my $i = $_[N];
                               real_foo($i);
```

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+    OUT int i            sub foo {
+                              my $i;
+                              real_foo(\$i);
+                              $_[N] = $i;
+
+    OUTLIST int i        sub foo {
```
```patch
                               $_[N] = $i;
+                         }
 
     OUTLIST int i        sub foo {
```

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+bookkeeping is usually handled for you automatically. Once you start
+handling this yourself, then there are some specific considerations.
+
+Functions which create a new SV, such as as C<newSViv(i)>, return an SV
```
```perl
s/ as as / as /
```

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+fully-qualified Perl name.
+
+If the name includes the package separator, C<::>, then it is treated as
+as a C++ method declaration, and various extra bits of processing take
```
```perl
s/ as as / as /  # split across lines
```

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+Usually you could dispense with C<PREINIT> by just wrapping the code in
+C<CODE> blocks in braces, but it may be necessary if the ordering of the
+variable initialisations is sensitive, e.g. if it affected by some
```
```perl
s/ if it affected / if it is affected /
```

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+consists solely of the C<INIT> keyword described below, plus the keywords
+described in L<XSUB Generic Keywords> and L<Sharing XSUB bodies>, plus the
+L<C_ARGS|/The C_ARGS: Keyword>, L<INTERFACE|/The INTERFACE: Keyword> and
+L<INTERFACE_MACRO|/The INTERFACE_MACRO: Keyword> keywords.
```
`INTERFACE` and `INTERFACE_MACRO` *are* described in “Sharing XSUB bodies”, so listing them is redundant.

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+=head3 The INIT: Keyword
+
+The C<INIT> keyword allows arbitrary initialisation code to inserted after
```
```perl
s/ to inserted / to be inserted /
```

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
 
-Parameters preceded by C<OUTLIST> keyword do not appear in the usage
-signature of the generated Perl function.
+This could be regarded as a bug.
```
Does “bug” imply that it might get fixed, and the escaped syntax `OVERLOAD: \"\"` might become illegal in future?

This feels similar to an *explicit* `OUTPUT: RETVAL` being required for non-autocall usage, which is described as “probably a bad design decision, but we're stuck with it now” [in the section on `CODE` above](https://github.com/Perl/perl5/blob/9afbfdafc5e26a350cb4c430f5d881e35e598f43/dist/ExtUtils-ParseXS/lib/perlxs.pod#the-code-keyword).

Should these two cases be described with similar language, or is there a real difference here with regards to possible future changes?

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+object); and third, a swap flag. See L<overload> for the full details
+of how these functions will be called, with what arguments. Note that
+C<swap> can in fact be undef in addition to false, to indicate an assign
+overload such as C<+=>.
```
To detect the difference between `swap` being false and undef, you’d need to declare it as `SV*` and use `SvOK()` and `SvTRUE()`, right? Should this be pointed out, or is it obvious enough?

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
-to make it a string is recommended if long version numbers are used.
+See L<INTERFACE|/The INTERFACE: Keyword> below for an alternative to
+C<ALIAS> which is more suited for autocall. Note that C<ALIAS> should not
+be used together with either of C<INTERFACE> or C<ATTRS>.
```
For symmetry with the section on `OVERLOAD`:
```patch
-be used together with either of C<INTERFACE> or C<ATTRS>.
+be used together with either of C<ATTRS>, C<INTERFACE>, or C<OVERLOAD>.
```

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+Similarly, common C and Perl unsigned types map to C<T_UV>, and values
+are converted back and forth via C<(UV)> casts. A few unsigned types such
+as C<I16> and C<U32> are instead mapped to C<T_U_SHORT> and C<T_U_LONG> XS
```
```perl
s/I16/U16/
```

johannessen commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+There are several variations on this basic pattern, which are explained in
+the following subsections.
+
+=head3 Fully-qualified type names and Perl objects
+
+    Foo::Bar
+    foo(Foo::Bar self, ...)
+
+Normally the type of an XUB's parameter or return value is a valid C type,
+such as C<"char *">. However you can also use Perl package names. When a
```
What happens when such a parameter receives a Perl object blessed to a different package? Is `@ISA` a factor at all?

I don’t know the answer, but feel like this should be addressed (briefly) somewhere in this document. For example the section [The OVERLOAD: Keyword](https://github.com/Perl/perl5/blob/9afbfdafc5e26a350cb4c430f5d881e35e598f43/dist/ExtUtils-ParseXS/lib/perlxs.pod#the-overload-keyword) makes it sound like there *is* an automatic type check, but it remains unclear whether it's just for the `svtype` or if the package is considered in some way:

> will croak with an `Expected foo to be of type My::Num, got scalar` error

tonycoz replied:
It depends on the typemap, the commonly used `T_PTROBJ` follows `@ISA`.

`OVERLOAD` just does `use overload 'op' => \&xs_sub`.

johannessen replied:
Ah, of course. Yes. Thank you for reminding me.

I suggest adding a [direct link for the T_PTROBJ documentation](https://metacpan.org/dist/ExtUtils-ParseXS/view/lib/perlxstypemap.pod#T_PTROBJ), for example like this:

```patch
@@ -4135,10 +4135,10 @@
 argument, until finally some sort of destroy function frees the handle and
 its data. The C<T_PTROBJ> typemap is one common method for mapping Perl
-objects to such C library handles. Behind the scenes, it uses blessed
-scalar objects with the scalar's integer value set to the address of the
-handle. The C<INPUT> code template of the C<T_PTROBJ> typemap retrieves the
-pointer from the scalar object referred to by a passed RV argument, while
-the C<OUTPUT> template creates a new blessed RV-to-SV with the handle
-address stored in it.
+objects to such C library handles; see L<perlxstypemap/T_PTROBJ>. Behind the
+scenes, it uses blessed scalar objects with the scalar's integer value set
+to the address of the handle. The C<INPUT> code template of the C<T_PTROBJ>
+typemap runs type checks and retrieves the pointer from the scalar object
+referred to by a passed RV argument, while the C<OUTPUT> template creates a
+new blessed RV-to-SV with the handle address stored in it.
 
 For the purposes of an example, we'll create here a minimal example C
```

johannessen reviewed (changes requested):
I left some comments, mostly about minor things.

Overall, I found the document to be well structured and the prose to be easily readable, in spite of the fairly complex topic.

Leont commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+    short
+    baz(int a, char *b = "")
+      PREINIT:
```
Should we still use `PREINIT`? I see it's use-case in C89 for declarations, but in C99 it can be confusing: mainly because it runs before argument handling.

Leont commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+completely within the signature (i.e. which don't use an C<INPUT> section
+to specify their type).
+
+Perls before 5.36 used C89 compiler semantics, which didn't allow variable
```
I don't think this is entirely true. We used C89 it for perl itself, but not necessarily for CPAN modules.

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+    /* For efficiency, always define PERL_NO_GET_CONTEXT: not enabled by
+     * default for backwards compatibility. For details, see
+     * L<perlguts|perlguts/
```
Is this intended to be a link to other documentation? Because this is in a code block, it won't be treated that way by pod parsers

tonycoz commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
-in your XS code, the XSUBs following this line will not be declared C<static>.
-You can later disable this with
+    int
+    X::Y::new(int i)
```
Shouldn't new() return an `X::Y *`?

tonycoz commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+consists of calling the PP function associated with the current OP
+(C<PL_op>) and then updating it, typically to C<< PL_op->op_next >>. In
+particular, the C<OP_ENTERSUB> performs (or at least starts) a function
+call.
```
99%* of XS code doesn't touch OPs, I don't think it needs to be the first heading, or even mentioned at all here.

\* I made this number up

tonycoz commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+=head3 The SV scalar value structure
+
+As mentioned above, almost runtime data within the perl interpreter is
```
I'm not sure it's good to go into so much detail on the internal structure of SVs here.

I expect nearly all XS code is fine with newSViv()/sv_setiv()/SvIV() etc rather than looking at any flags beyond SvOK() and SvUTF8().

johannessen replied:
I think this entire section on SVs is put together well and parts of it are quite important. It’s not obvious to me what could be cut. Could you be more specific?

The flags in particular might not be useful information for everybody, but explaining the general principle of how an SV works I feel is crucial, and the flags are a big part of that. FWIW, reading that part made me recognise two bugs in one of my modules. I think this section is fine.

tonycoz commented:
I'm done for now.

iabyn commented:
On Mon, Oct 20, 2025 at 07:58:13PM -0700, Tony Cook wrote:
> > +=head3 Perl OPs
...
> 99%* of XS code doesn't touch OPs, I don't think it needs to be the
> first heading, or even mentioned at all here.

I mentioned OPs mainly as a way of introducing OP_ENTERSUB, which I needed
to do to explain how XSUBs get called.

Also, I wanted to give readers a bit of background knowledge abut how the
interpreter works, to make sense of things when debugging.

I'll think about this some more.

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+Note that much of the L<data passing overview|/Overview of how data is
+passed to and from an XSUB> part of this document is just a summary of
+the parts of of perlguts which are most relevant to writing XS code,
```
nit: `s/of of/of/`

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+them it should be done in a way which doesn't increase its reference
+count. For example, this modifies C<sv> to be a reference to a
+newly-created SV holding an integer value, i.e. the perl equivalent of
+C<$sv =\99>:
```
nit: A space after the '=' would be good so it looks like `$sv = \99`

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+reference count to it), then if the code croaks, the SV on the stack will
+leak. To avoid this, there is a separate I<temps> stack in the Perl
+interpreter. Items on this stack I<are> reference counted. Typically the
+temps stack is reset at start of each statement, back to some particular
```
nit: at start -> at the start

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+    SV *sv_abc = newSVpvn_flags("abc", 3, SVs_TEMP);
+
+Many OPs have an SV attached to them called a I<PADTMP>. This SV has a
+lifetime which is the same as the sub which the OP is a part of, and
```
"This SV has a lifetime which is the same as the sub which the OP is a part of..."

This feels like "As long as the sub is defined..." but I *think* this means "the duration of the subroutine call"?

tonycoz replied:
No, it lives as long as the sub (don't use it to return references), it saves the OP the cost of creating a new SV each time it executes.

wolfsage replied:
Perhaps this could be clarified then, but also perhaps I'm being thick :)

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+=head3 The SV scalar value structure
+
+As mentioned above, almost runtime data within the perl interpreter is
```
nit: almost runtime -> almost all runtime

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+Simple SVs may not have a body. Undefined values typically don't have one.
+Also, some IV, NV, and RV values are stored directly in the payload field,
+with the body pointer pointing back to the head with a suitable offset so
```
...with a suitable offset so that the SV appears to have a body with an IV or whatever at a suitable offset" is weirdly redundant. Can this be clarified at all?

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+which calculates the integer value from its current string value. The
+effect of this call is to update the SV's type and body to C<SVt_PVIV>
+which is capable of holding both a string I<and> and integer value, and
+then to set the C<SVf_IOK> flag in addition to the C<SVf_POK>.
```
nit: "in addition to the C<SVf_POK> flag"

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+Note that you should never directly access fields using macros like
+C<SvIVX()> (the C<X> implies direct) I<unless> you have just tested for
+C<SvIOK()>. In general, always use macros such as C<SvIV()>, which will do
```
This almost implies SvIOK() says any Sv*X is going to work. Maybe it could be reworded...

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+    STRLEN len;
+    char *pv = SvPV(sv, len);
+
+(and variants) which both retrieves a string pointer and sets C<len> to
```
(and variants) reads really weird here. Maybe....

```
The equivalent of C<SvIV()> for strings (and variants) is:
```

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+An important thing to note is that AVs and HVs are never directly pushed
+onto the stack when calling and returning from subroutines and XSUBs.
+Instead where necessary, references to them are pushed. You will likely
```
Might be good to reiterate this means an RV and/or how to do it)

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+in C code (e.g. not in an XSUB signature). More generally, the XS parser
+doesn't understand C syntax or semantics; it just uses crude regexes to
+parse the XS file. For example in an XSUB declaration like
+
```
This example doesn't really explain how the declaration is parsed. Does it get it right? Or horribly wrong?

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+parameters (including any default values). While it is modelled on C
+syntax, it is actually XS syntax (so for example C</* ... */> isn't
+recognised). The return type and name must both start on column one,
+although The XS parser actually allows both to be on the same line, such
```
nit: The -> the

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+go in places where a variable can be declared, and typically declare one
+or more variables and possibly their initialisations.
+
+C<dXSARGS> pops one index off the mark stack and sets up some auto
```
dVAR isn't mentioned at all in this document beyond the example above. Should it be before dXSARGS here?

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+to specify their type).
+
+Perls before 5.36 used C89 compiler semantics, which didn't allow variable
+declarations after statements. To work round this, the C<PREINIT> keyword
```
nit: work round -> work around

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+The C<XSRETURN_UNDEF> macro causes the address of the special Perl SV
+C<PL_sv_undef>  to be stored at C<ST(0)> (this is the same value that the
+Perl function C<undef> returns), and then to immediately return.
```
nit: perhaps instead of "and then to immediately return" ->  "and then immediately returns"

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+But in addition, for I<any> C<$arg = ...> template (not just the template
+for C<SV*>), the XS compiler makes one further assumption: that the
+expression to the right of the assign evaluates to an SV with a reference
+count count I<one too high>, and so in addition, the XS compiler emits:
```
nit: count count

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+However, if the SV comes from elsewhere, for example via a Perl array
+lookup, then its reference count doesn't need to be adjusted, and so the
+mortalising will cause it to be prematurely freed. In this case, you need
+to artificially increase the SV's reference count.
```
Could be nice to say "...artificially increase the SV's reference count using SvREFCNT_inc as seen blow"

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+Sometimes you want to return a non-scalar SV, such as an AV, HV or CV.
+However, these aren't allowed directly on the argument stack. You are
+supposed instead to return a I<reference> to the AV: a bit like a Perl sub
```
nit: supposed instead to -> supposed to instead

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+which just returns a scalar). In this case you have to tell the C compiler
+that C<SVREF> is just another name for C<SV*>:
+
+    typedef SV *SVREF;
```
That's weird. Why is this necessary (when no other types have to be declared in this way)

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+the input typemap entry for C<AV*> automatically takes care of
+dereferencing the argument and croaking if it's not actually a reference.
+The C<PUSHs()> macro simply pushes an SV onto the stack, without any
+mortalising or copying. Any "holes" in the array are filled with undefs.
```
I'm mildly surprised the above section didn't touch on the XPUSH* macros that extend the stack for you

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+    XSLoader::load(__PACKAGE__, $VERSION);
+
+This causes the F<Bar.so> or F<Bar.dll> file to be dynamically linked in
+and then the C<boot_Foo__Bar()> function called. This boilerplate code is
```
nit: function called -> function to be called

wolfsage reviewed (commented):
(more to come, don't want this to get lost)

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+use C<TYPEMAP:> to individually override specific C<TYPEMAP>, C<INPUT>, or
+C<OUTPUT> entries in the system typemap. In general, typemap changes
+affect any subsequent XSUBs within the file, until further updates.  Note
+however that due a quirk in parsing, it is possible for a C<TYPEMAP:>
```
nit: due a quirk -> due to a quirk

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+C<OUTPUT> entries in the system typemap. In general, typemap changes
+affect any subsequent XSUBs within the file, until further updates.  Note
+however that due a quirk in parsing, it is possible for a C<TYPEMAP:>
+entry immediately I<after> an XSUB to affect that XSUB.
```
Wow! Can we provide more info here or a "how to avoid this"? Otherwise I wouldn't risk using it...

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+=back
+
+The most recently-applied entries take precedence, so for example you can
```
nit: recently-applied -> recently applied

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+      CODE:
+        ...
+
+Generally there no reason to use the old style any more, apart from a few
```
nit: there is no reason

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+the keyword). For the first, it may appear anywhere in the input part or
+the XSUB. For the latter, it may appear anywhere in file scope, but due to
+a long-standing parser bug, the keyword's state is reset at the start of
+each XSUB, so it will only have any effect if appears just before a XSUB
```
nit: if appears -> if it appears

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
-semicolon is allowed after the argument list, as in
+Only one of these keywords may appear in this part, and at most once; and
+no other keywords are recognised in this part (although such keywords
+could be instead be processed in the tail or head of the preceding and
```
nit: could be instead be -> could instead be?

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
-          XSRETURN(1);
+=head3 Auto-calling a C function
+
+In the absence of any explicit main body code via C<CODE> or C<PPCODE>,
```
or NOT_IMPLEMENTED_YET !

wolfsage reviewed (commented):
Alright that's enough for one night, more later

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+The C<POSTCALL> keyword allows a block of code to be inserted directly after
+any autocall or C<CODE>/C<PPCODE> code block (although it's really only of
+use with autocall). It's typically used for cleaning up the return value
```
I'm confused, up above we say:

```
The C<PPCODE> keyword must be the last keyword in the XSUB. 
```

If PPCODE must be the last keyword in the XSUB, how can it have a POSTCALL?

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+and C<arg2> will have set magic, while C<arg1> won't. The C<SETMAGIC>
+setting remains in force until another C<SETMAGIC>, or notionally until
+the end of the current C<OUTPUT> block. In fact the current setting will
+carry over into any further C<OUTPUT> declarations within in the same
```
nit: within in the same -> within the same

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+design flaw and should probably be avoided. Since 5.16.0 it's been
+possible to have locally defined typemaps using the L<TYPEMAP|/The
+TYPEMAP: Keyword> keyword which is probably a better way to modify how
+values are returned.
```
Would it be useful to show the equivalent typemap here?

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+
+The original C<mynum_add()> method which, via aliasing, handled all four
+of the arithmetic operations, is now split into four separate XSUBs, since
+C<ALIAS> and C<OVERLOAD> doesn't mix.
```
nit: "don't mix"

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+the C<printf> above). The ellipsis is added to ignore the extra two
+arguments passed to an overload method.
+
+The original C<mynum_add()> method which, via aliasing, handled all four
```
This is mildly confusing since "the original mynum_add..." is way below this section and hasn't been seen yet

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+of the arithmetic operations, is now split into four separate XSUBs, since
+C<ALIAS> and C<OVERLOAD> doesn't mix.
+
+The main change to each arithmetic XSUB part from adding the C<OVERLOAD>
```
nit should "part from adding" be "apart from adding" ?

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+      ATTRS: lvalue
+      PPCODE:
+        # return $Foo::Bar::DEBUG, creating it if not already present:
+        PUSHs(GvSV(gv_fetchpvs("Foo::Bar::DEBUG", GV_ADD, SVt_IV)));
```
Woah, why doesn't this need to extend the stack first? If it doesn't an explanatory comment might help others with the same question

tonycoz replied:
There's always room for one element on the stack (the spot the CV was popped from in pp_entersub).

I experimented with requiring the extend for 1 argument returns in the HWM check: it broke a lot of things.

wolfsage replied:
Ah, I recall coming across that before! Right... I think this could definitely use a comment then

iabyn committed:
perlxs: tweaks following code review

The many commits in this branch have completely rewritten perlxs.pod.
This commit applies all the minor tweaks suggested by reviewers.

(It was far too much like hard work to try and update each individual
commit with the various changes.)

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+where typemaps can sometimes be of use (and sometimes aren't).
+
+Standard signed C int types such as C<int>, C<long> and C<short>, are all
+mapped to to the C<T_IV> XS type. Integer-like Perl types such C<IV> and
```
"such C<IV>" -> "such as C<IV>"

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
-	$status = x_gettime( $timep, $host );
+Note that, as a very special case, the XS compiler translates the XS
+typemap name using C<s/OBJ$/REF/> when looking up INPUT typemap entries
+for an XSUB named C<DESTROY>. So for such subs, the C<T_PTRREF> typemap
```
This is the first time T_PTRREF is encountered in this document. Maybe a word on it would be useful otherwise I'm left wondering "what is it?"

wolfsage replied:
(even if it's just a note to go read perlxstypemap...)

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+    void
+    Paint::color::DESTROY()
+
+In the C part of the XS file (or this case, the C++ part), a trivial
```
"or this case" -> "or in this case"

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+an I<exact> copy of C<T_PTRREF>. This is needed because, as an
+optimisation, the XS parser automatically renames an INPUT typemap using
+C<s/OBJ$/REF/> if the name of the XSUB is C<DESTROY>, on the grounds that
+it's not necessary to to check that the object is the right class.
```
"to to" -> "to"

wolfsage commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4706 @@
+        if (index > MY_CXT.count)
+            croak("There are only %d blind mice.", MY_CXT.count);
+        else
             RETVAL = MY_CXT.name[index - 1];
```
"index" is a weird name for this arg since you're really saying "give me the first, second, or third mouse.."

And if you pass 0 we end up reading ```[index - 1] !```

wolfsage reviewed (commented):
Final review done! Few more small things found.

iabyn committed:
perlxs.pod: standardise version numbers in the doc

After the big rewrite, various bits of text which describe when a
particular feature was introduced or changed have ended up using a
random mixture of Perl, ParseXS and xsubpp version numbers.

This commit standardises on xsubpp version numbers. These are mostly
the same as ParseXS, but this handles the cases before ParseXS was
split off from xsubpp. Perl versions suffer from not exactly matching
when an xsubpp version number was incremented, and not matching what is
used by the "REQUIRE:" keyword.

This commit also adds a short new section which tries to explain how the
three sets of version numbers are related.

tonycoz commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -3628,7 +3694,9 @@ to be rewritten more cleanly as:
-        # return $Foo::Bar::DEBUG, creating it if not already present:
+        # return $Foo::Bar::DEBUG, creating it if not already present
+        # (NB: XPUSHs() not needed here as the stack always has one
+        allocated slot available when an XSUB is called):
```
This last line of the comment should have a `#`

iabyn replied:
fix pushed

Leont commented:
> As a general rule, if I didn't reply to a comment, it means I implicitly accepted the point.

Please mark comments as resolved when you're done with them, that cleans up this thread immensely for anyone reading it. Now it isn't even rendering everything because there are so many of them.

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+
+XS has extensive support for the first role, and makes writing the second
+need less boilerplate code. This document doesn't cover the third role,
+which often requires extensive knowledge of the Perl interpreter's
+internals.
```
Are you calling me out here? I feel called out ;)

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+
+(This next paragraph is definitely only for background on debugging.)
+
+An C<OP> is is a data structure within the perl interpreter. It is used to
```
Duplicate "is is"
```suggestion
An C<OP> is a data structure within the perl interpreter. It is used to
```

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+The various OPs executed by the Perl interpreter up until the function is
+called will: push a mark indicating the start of a new argument stack
+frame; push an SV containing the integer value 1; push the SV currently
+bound to the variable C<$x>; and push the C<*foo> typeglob. Then the
```
I think "bound to" is likely to confuse folks not from a Lisp/Scheme background. Maybe "the SV currently representing the variable C<$x>" ?

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+For a normal Perl subroutine call, C<pp_entersub()> will then: pop the
+topmost mark off the mark stack; pop the SV pointers between that mark and
+the top of the stack and store them in C<@_>; then set C<PL_op> to the
+first  OP pointed to by the C<&foo> CV. Those OPs will then be run by the
```
Double space "first  OP".

Also, I see some documentation I'm going to have to alter if I get the CVf_NOSNAIL flag working :)

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+
+As mentioned above, almost all runtime data within the perl interpreter is
+stored in an SV (scalar value) structure. The head of an SV structure
+consists of three or four words: a reference count; a type and flags; a
```
I think "words" is possibly not a word [pardon the phrasing] that would make sense to folks outside of a machine-level programming background. I'd say "fields", just relating to the members of the C struct.

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+    STRLEN len;
+    char *pv = SvPV(sv, len);
+
+which both retrieves a string pointer and sets C<len> to its length. Note
```
It might help casual readers to explicitly point out that since `SvPV` is a macro, it can do the evil trickery of modifying `len` directly, so you don't have to pass in `&len`.

leonerd reviewed (commented):
Wow, a much bigger (and better) document than previous. So much so I've only had time to review about the top 20% so far but I wanted to commit these comments before they get lost, and to say I will review the rest in the next day or so.

Overall so far looking very much better, here's just a few tiny nits.

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+      OUTPUT:
+        RETVAL
+
+Finally, note that some very old  (pre-1996) XS documentation suggested
```
Another double-space on this line

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+      PPCODE:
+        {
+            int i;
+            int max_ix = AvFILL(av);
```
Are we intending these code examples still run on (much) older perls? For a while now we've had `av_count()` which returns a count, so you don't have to do the more awkward `+ 1` logic elsewhere. But that wouldn't work before it was added - 87306e0674d in time for 5.34.

iabyn replied:
In general I'm trying to make everything compatible as far back as 5.8.x.

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+limited use and are typically only used to mimic the behaviour of Perl
+builtins. For example there is no way to implement a C<push @a, ...;>
+style function without a way of telling the Perl interpreter not to
+flatten C<@a>. Outside of these  narrow uses, it is generally a mistake to
```
Double-space

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+
+Note however that, due to a quirk in parsing, it is possible for a
+C<TYPEMAP:> block which comes I<immediately after> an XSUB to affect any
+entries used by that XSUB, as if the  block had appeared just before the
```
Double-space

Also: Oh gosh I can't imagine the crazy amount of parsing logic that results in this particular quirk. ;)

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+An XSUB definition consists of a declaration (typically two lines),
+followed by an optional body. The declaration specifies the XSUB's name,
+parameters and return type. The body consists of sections started by
+keywords, which may specify how its parameters and any any return value
```
Duplicated "any any"

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+
+In fact all it does is extract the text between the C<(...)> and split on
+commas, while having enough intelligence to ignore commas and a closing
+parenthesis within a double-quoted string. Once each parameter declaration
```
Specifically double-quoted? What about `char comma = ','` - is that handled?

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+the common C<T_PTROBJ> typemap type.
+
+Note that any check on whether the passed Perl object is of the correct
+class is down to to the implementation in the particular typemap: for
```
Duplicated "to to"

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+    foo(int i, char *s = "abc")
+
+    int
+    bar(int i, int j = i + ')', char *s = "abc,)")
```
Aha, this example suggests the parser can handle double- and single-quoted strings. Does the wording earlier need to say that?

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+will have to access any extra arguments manually, making use of the
+C<ST(n)> macro to access the nth item on the stack (counting from 0), and
+the C<items> variable, which indicates the total number of passed
+arguments, including any fixed arguments.
```
Might be worth clarifying here, that `ST(0)` counts the first argument overall, so the CODE body will have to account for counting its own initial fixed arguments. I see the example below does that, but perhaps useful to add more words here?

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+
+Note that the L<PPCODE|/The PPCODE: Keyword> keyword, in comparison to
+C<CODE>, resets the local copy of the argument stack pointer, and relies
+on the coder  to place any return values on the stack. The example above
```
Double-space

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+    void
+    triple(...)
+      PPCODE:
+        SP += items;
```
I'm not sure I follow why this line is here. The `ST()` macro used below doesn't use `SP`, so it won't have affect there. I suppose it could be useful for any invoked Perl code implied within (e.g. overload or magic on the SVs), but for that to have affect you'd have to `PUTBACK` first.
You actually address all those details in the description further down, just I think this code is missing it.

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
-module to enable and disable prototypes for different parts of the module.
-Note that B<xsubpp> will nag you if you don't explicitly enable or disable
-prototypes, with:
+This could be regarded as an implementation bug, but we're stuck with it
+now.
```
If ever there was one line that summarised the whole of the internals documentation, this could be it ;)

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+details of how to set the fallback behaviour for the current package.
 
-    PROTOTYPES: DISABLE
+Note that C<OVERLOAD> shouldn't be mixed with the L<ALIAS|/The ALIAS:
+Keyword> keyword; the value of C<ix> will be undefined for any overload
+method call.
```
Aside: Is this an implementation bug that we should consider fixing?

iabyn replied:
it's on my list of things to address.

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+information stored in the CV indicates which C library function should be
+autocalled.
+
+Finally, there is the C<CASE> keyword,  which allows the whole body of an
```
Double-space

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
-and the function pointer.
+The alias name can be either a simple function name or can include a
+package name. The alias value to the right of the C<=> may be either a
+literal positive integer or a word (which is expected to be a CPP define).
```
... or `enum` constant. I've done that several times.

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+    $color->set_blue(0x80);
+    printf "blue=%d\n", $color->blue();   # prints 128
 
 =head2 Safely Storing Static Data in XS
```
I believe this entire section is talking about *mutable* static data, yes? I've seen (and written) many cases of `static const` data structures being used all over .xs files. As they're `const` and set up purely at compiletime, there's no issues with mutations across threads.

Should there be perhaps some wording on that topic?

Leont replied:
Yeah, though it would also include a `static SV* const` and similar.

leonerd commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -2230,22 +4946,24 @@ this model, the less likely conflicts will occur.
 =head1 AUTHOR
 
 Originally written by Dean Roehrich <F<roehrich@cray.com>>.
+Completely rewritten in 2025.
```
Feel free to add your name here of course :)

leonerd reviewed (commented):
I've got around to reading and reviewing the entire rest.
I admit I somewhat skimmed over the C++ related parts as I don't know the technical details here, though I have read over all the wording and style at least.

Overall a few small comments but generally this is looking much better indeed.

Leont commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+(including most parameter and return value processing) within an C<{
+ENTER;> and C<LEAVE; }> pair. This has the effect of clearing any
+accumulated savestack entries at the end of the code body. If disabled,
+then the savestack will usually be cleared by the caller anyway, so this
```
"The caller" here means leavesub, right? I don't really understand when one would ever need this. `MULTICALL` shenanigans?

iabyn replied:
"caller" in this case is technically pp_entersub(); leavesub is the last op of a *perl* sub. But I have no idea why SCOPE was added; it was some time before 5.004 in 1996, and the p5p thread is just a bare 'here's a patch to add the SCOPE' keyword, with no indication of what sort of problem it was written to address. MULTICALL was added much later.

Leont commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -1,2065 +1,4803 @@
+contains a code comment like C</* SCOPE */>, then scoping will be
+automatically enabled for any XSUB which uses that typemap entry. This
+currently only works for parameters whose type is specified using
+old-style C<INPUT> lines rather than an ANSI-style declaration, i.e. not
```
Treating the two argument styles differently sounds to me like a bug we should fix instead of something we should document.

iabyn replied:
It's definitely a bug. It's on my (huge) list of XS parser things to fix; but for now, lets document that's it's an issue to be aware of.

iabyn committed:
perlxs: more tweaks following code review

The many commits in this branch have completely rewritten perlxs.pod.
This commit applies a second set of minor tweaks suggested by reviewers.

iabyn committed:
perldelta entry for perlxs.pod rewrite

khwilliamson commented:
On 12/17/25 05:57, David Mitchell wrote:
> Merged #23795 <https://github.com/Perl/perl5/pull/23795> into blead.

iabyn++

> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ 
> Perl/perl5/pull/23795#event-21616792169>, or unsubscribe <https:// 
> github.com/notifications/unsubscribe-auth/ 
> AAA2DH7N4BMRGYXM5J2V5TT4CFHKHAVCNFSM6AAAAACIDDHZ5KVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMRRGYYTMNZZGIYTMOI>.
> You are receiving this because you are subscribed to this thread.Message 
> ID: ***@***.***>
>
